packages/@react-aria/menu/src/useSubMenuTrigger.ts

```diff
@@ -0,0 +1,244 @@
+/*
+ * Copyright 2023 Adobe. All rights reserved.
+ * This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under
+ * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ * OF ANY KIND, either express or implied. See the License for the specific language
+ * governing permissions and limitations under the License.
+ */
+
+
+import {AriaMenuItemProps} from './useMenuItem';
+import {AriaMenuOptions} from './useMenu';
+import type {AriaPopoverProps} from '@react-aria/overlays';
+import {FocusableElement, FocusStrategy, PressEvent, Node as RSNode} from '@react-types/shared';
+import {RefObject, useCallback, useRef} from 'react';
+import type {SubMenuTriggerState} from '@react-stately/menu';
+import {useEffectEvent, useId, useLayoutEffect} from '@react-aria/utils';
+import {useLocale} from '@react-aria/i18n';
+
+export interface AriaSubMenuTriggerProps {
+  /** An object representing the submenu trigger menu item. Contains all the relevant information that makes up the menu item. */
+  node: RSNode<unknown>,
+  /** Whether the submenu trigger is disabled. */
+  isDisabled?: boolean,
+  // TODO: naming. Also talk about if this should be customizable in this hook or if it belongs somewhere else
+  /** Type of the submenu being rendered. */
+  subMenuType?: 'dialog' | 'menu',
+  /** Ref of the menu that contains the submenu trigger. */
+  parentMenuRef: RefObject<HTMLElement>,
+  /** Ref of the submenu opened by the submenu trigger. */
+  subMenuRef: RefObject<HTMLElement>
+}
+
+export interface SubMenuTriggerAria<T> {
+  subMenuTriggerProps: AriaMenuItemProps,
+  subMenuProps: AriaMenuOptions<T>,
+  popoverProps: Pick<AriaPopoverProps, 'isNonModal'>,
+  // TODO: descriptions
+  // TODO: perhaps cherry pick these props from existing types, comes from several different types though
+  overlayProps:
+  {
+    onExit: () => void,
+    disableFocusManagement: boolean,
+    shouldCloseOnInteractOutside: (element: Element) => boolean
+  }
+}
+
+// TODO description
+// TODO: debatable if we should have a useSubMenu hook for the submenu keyboard handlers. Feels better to have it here so we don't need to ferry around
+// things like parentMenuRef and the open timeout canceling since we centralized them here
+export function UNSTABLE_useSubMenuTrigger<T>(props: AriaSubMenuTriggerProps, state: SubMenuTriggerState, ref: RefObject<FocusableElement>): SubMenuTriggerAria<T> {
```

Comment:
API nit: I think we should use `Submenu` and not `SubMenu` based on usage across the web.

Reply:
on second thought, the current api matches v2's naming so in the interest of ease of upgrading I'm going to keep as is.